packages/cozy-harvest-lib/src/services/budget-insight.js

```diff
@@ -0,0 +1,129 @@
+import get from 'lodash/get'
+import omit from 'lodash/omit'
+import merge from 'lodash/merge'
+import clone from 'lodash/clone'
+import defaults from 'lodash/defaults'
+
+import { waitForRealtimeResult } from './jobUtils'
+import {
+  createBIConnection,
+  updateBIConnection,
+  getBIConfig,
+  getBIModeFromCurrentLocation,
+  isBudgetInsightConnector
+} from './biUtils'
+import assert from '../assert'
+import { mkConnAuth } from 'cozy-bi-auth'
+
+const getBIConnectionIdFromAccount = account =>
+  get(account, 'data.auth.bi.connId')
+
+const throwWrappedError = (originalError, namespace) => {
+  const error = new Error(`${namespace} failed (${originalError.message})`)
+  error.original = originalError
+  throw error
+}
+
+const createTemporaryToken = async ({ client, account, konnector }) => {
+  try {
+    assert(
+      account && account._id,
+      'createTemporaryToken: Invalid account for createTemporaryToken'
+    )
+    assert(konnector.slug, 'createTemporaryToken: No konnector slug')
+    const jobResponse = await client.stackClient.jobs.create('konnector', {
+      mode: 'getTemporaryToken',
+      konnector: konnector.slug,
+      account: account._id
+    })
+    const event = await waitForRealtimeResult(
+      client,
+      jobResponse.data.attributes
+    )
+    return event.data.code
+  } catch (e) {
+    throwWrappedError(e, 'createTemporaryToken')
+  }
+}
+
+/**
+ * Ensures a BI connection is ready
+ *
+ *  - Calls the getBITemporaryToken mode of the konnector
+ *
+ * Then:
+ * If no BI connection is present in the io.cozy.accounts:
+ *  - Creates the BI connection using user supplied data
+ * If a BI connection exists:
+ *  - Updates the BI connection using user supplied data
+ *
+ * @param  {Client} options.client
+ * @param  {io.cozy.accounts} options.account
+ * @param  {io.cozy.konnectors} options.konnector
+ * @return {BIConnection}
+ */
+export const createOrUpdateBIConnection = async ({
+  account,
+  client,
+  konnector
+}) => {
+  try {
+    const mode = getBIModeFromCurrentLocation()
```

Comment:
This will not work on mobile apps since the location on a mobile app is not the URL of the cozy. Is that a problem?

Reply:
You're right, I will do that from the client uri !